### PR TITLE
JBPM-9327 Add excute shell script workitem to workitem repository

### PR DESCRIPTION
**Pull request by sudhishmk**
**Thank you for submitting this pull request**

**JIRA**: _(please edit the JIRA link if it exists)_ 

[link](https://issues.redhat.com/browse/JBPM-9327)

<pre>
How to retest a PR or trigger a specific build:

* <b>a pull request</b> please add comment: <b>Jenkins retest this</b>
 
* <b>a full downstream build</b> please add comment: <b>Jenkins run fdb</b>
  
* <b>a compile downstream build</b> please  add comment: <b>Jenkins run cdb</b>

* <b>a full production downstream build</b> please add comment: <b>Jenkins execute product fdb</b>

* <b>an upstream build</b> please add comment: <b>Jenkins run upstream</b>
</pre>

### CODE REVIEW

**Comment by sudhishmk** [comment]
Can one of the admins review this PR?

**Comment by krisv** [comment]
ok to test

File: exec-workitem/src/main/java/org/jbpm/process/workitem/exec/ExecShellScriptWorkItemHandler.java [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+ * WorkItemHandler that is capable of executing shell script file
+ * <ul>
+ * <li>ShellScriptLocation - Absolute or relative path of shell script file - mandatory</li>
+ * </ul>
+ */
+
+
+@Wid(widfile = "ExecShellScriptDefinitions.wid", name = "ExecShellScript",
+        displayName = "ExecShellScript",
+        defaultHandler = "mvel: new org.jbpm.process.workitem.exec.ExecShellScriptWorkItemHandler()",
```
Comment (tsurdilo) [inline exec-workitem/src/main/java/org/jbpm/process/workitem/exec/ExecShellScriptWorkItemHandler.java:50]: First question that comes to mind is - how is this different than the existing exec workitem  - https://github.com/kiegroup/jbpm-work-items/tree/master/exec-workitem

If they are the same thing, can the existing one be enhanced to do file support like what it seems you need here? Do we need two handlers??

File: exec-workitem/src/main/java/org/jbpm/process/workitem/exec/ExecShellScriptWorkItemHandler.java [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+        documentation = "exec-workitem/index.html",
+        category = "exec-workitem",
```
Comment (tsurdilo) [inline exec-workitem/src/main/java/org/jbpm/process/workitem/exec/ExecShellScriptWorkItemHandler.java:52]: This category should not be hard-coded, see the other handler java code in the same directory
 category = "${artifactId}",

File: exec-workitem/src/main/java/org/jbpm/process/workitem/exec/ExecShellScriptWorkItemHandler.java [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+ * <ul>
+ * <li>ShellScriptLocation - Absolute or relative path of shell script file - mandatory</li>
+ * </ul>
+ */
+
+
+@Wid(widfile = "ExecShellScriptDefinitions.wid", name = "ExecShellScript",
+        displayName = "ExecShellScript",
+        defaultHandler = "mvel: new org.jbpm.process.workitem.exec.ExecShellScriptWorkItemHandler()",
+        documentation = "exec-workitem/index.html",
```
Comment (tsurdilo) [inline exec-workitem/src/main/java/org/jbpm/process/workitem/exec/ExecShellScriptWorkItemHandler.java:51]: See the other workitem handler java code its like
documentation = "${artifactId}/index.html",

File: exec-workitem/src/main/java/org/jbpm/process/workitem/exec/ExecShellScriptWorkItemHandler.java [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+        category = "exec-workitem",
+        icon = "Exec.png",
```
Comment (tsurdilo) [inline exec-workitem/src/main/java/org/jbpm/process/workitem/exec/ExecShellScriptWorkItemHandler.java:53]: Icon should match the displayname defined...so in your case should be ExecShellScript.png

**Review by tsurdilo — CHANGES_REQUESTED** [review]
Just left some minor comments that should be addressed.

Overall the big question is why do we need a separate workitem handler for this? Can the existing one be enhanced maybe so users can pass this file param to it as well??

File: exec-workitem/src/main/java/org/jbpm/process/workitem/exec/ExecShellScriptWorkItemHandler.java [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+        parameters = {
+                @WidParameter(name = "ShellScriptLocation", required = true),
+        },
+        results = {
+                @WidResult(name = "Output")
+        },
+        mavenDepends = {
```
Comment (tsurdilo) [inline exec-workitem/src/main/java/org/jbpm/process/workitem/exec/ExecShellScriptWorkItemHandler.java:60]: Look at the other handler files in same dir, should be:
mavenDepends = {
                @WidMavenDepends(group = "${groupId}", artifact = "${artifactId}", version = "${version}")
        },

File: exec-workitem/src/main/java/org/jbpm/process/workitem/exec/ExecShellScriptWorkItemHandler.java [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+                @WidMavenDepends(group = "org.jbpm.contrib", artifact = "exec-workitem", version = "7.43.0-SNAPSHOT")
+        },
+        serviceInfo = @WidService(category = "Exec", description = "Execute a shell script",
```
Comment (tsurdilo) [inline exec-workitem/src/main/java/org/jbpm/process/workitem/exec/ExecShellScriptWorkItemHandler.java:63]: should be

 serviceInfo = @WidService(category = "${name}", description = "${description}",

File: exec-workitem/src/main/java/org/jbpm/process/workitem/exec/ExecShellScriptWorkItemHandler.java [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+    public void executeWorkItem(WorkItem workItem,
+                                WorkItemManager manager) {
+
+        try {
+        	
+
+            RequiredParameterValidator.validate(this.getClass(),
+                                                workItem);
+
+            String ShellScriptLocation = (String) workItem.getParameter("ShellScriptLocation");
```
[the remark this answers is not in the document]
Comment (sudhishmk) [reply exec-workitem/src/main/java/org/jbpm/process/workitem/exec/ExecShellScriptWorkItemHandler.java:82]: Oh, that was miss, will correct it. Thanks a lot for taking time to review the PR and providing suggestions. 

Let me review all your recommendation and will update in 1-2 days.

[the remark this answers is not in the document]
Comment (sudhishmk) [reply exec-workitem/src/main/java/org/jbpm/process/workitem/exec/ExecShellScriptWorkItemHandler.java]: Corrected.

File: exec-workitem/src/main/java/org/jbpm/process/workitem/exec/ExecShellScriptWorkItemHandler.java [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+            Map<String, Object> results = new HashMap<>();
+
+            Process process;
+            try {
+                
+                List<String> cmdList = new ArrayList<String>();
+                // adding command and args to the list
+                cmdList.add("sh");
+                cmdList.add(ShellScriptLocation);
+                ProcessBuilder processBuilder = new ProcessBuilder(cmdList);
```
[the remark this answers is not in the document]
Comment (sudhishmk) [reply exec-workitem/src/main/java/org/jbpm/process/workitem/exec/ExecShellScriptWorkItemHandler.java]: Yes, I had thought of it but wanted first check if this was valid enhancement before adding those capabilities. Thanks for the suggestion I have added it.

File: exec-workitem/src/main/java/org/jbpm/process/workitem/exec/ExecShellScriptWorkItemHandler.java [hunk 9]
```diff
@@ -0,0 +1,132 @@
...
+                process = processBuilder.start();
+                    
+                process.waitFor();
```
[the remark this answers is not in the document]
Comment (sudhishmk) [reply exec-workitem/src/main/java/org/jbpm/process/workitem/exec/ExecShellScriptWorkItemHandler.java]: Added

File: exec-workitem/src/main/java/org/jbpm/process/workitem/exec/ExecShellScriptWorkItemHandler.java [hunk 10]
```diff
@@ -0,0 +1,132 @@
...
+                String line; 
+                while((line = reader.readLine()) != null) { 
+                	output.add(line);
+                	logger.debug("Output line " + line);
+                	
+                } 
+            } catch (IOException e) {
+            	logger.error("Error executing the work item IO Exception: " + e.getMessage());
+            	handleException(e);
+            } catch (InterruptedException e) {
```
[the remark this answers is not in the document]
Comment (sudhishmk) [reply exec-workitem/src/main/java/org/jbpm/process/workitem/exec/ExecShellScriptWorkItemHandler.java]: Done

File: exec-workitem/src/main/java/org/jbpm/process/workitem/exec/ExecShellScriptWorkItemHandler.java (hunk 1 above)
Comment (sudhishmk) [reply exec-workitem/src/main/java/org/jbpm/process/workitem/exec/ExecShellScriptWorkItemHandler.java:54]: I had thought of enhancing the existing execute work item but felt that using processBuilder approach was better than commandLine executor for executing shell scripts. Enhancing existing workitem using processBuilder for executing file seemed not appropriate. I observed few other WIH handlers like GIT, twitter etc where separate handlers was created for separate operations. So felt that exceuting a command vs executing file would be separate operations, so decided with separate handler. It's the first time I am contributing to the open source repo, so I am not very sure about design details. Let me know if you have any recommendations.

File: exec-workitem/src/main/java/org/jbpm/process/workitem/exec/ExecShellScriptWorkItemHandler.java (hunk 2 above)
Comment (sudhishmk) [reply exec-workitem/src/main/java/org/jbpm/process/workitem/exec/ExecShellScriptWorkItemHandler.java]: updated

File: exec-workitem/src/main/java/org/jbpm/process/workitem/exec/ExecShellScriptWorkItemHandler.java (hunk 3 above)
Comment (sudhishmk) [reply exec-workitem/src/main/java/org/jbpm/process/workitem/exec/ExecShellScriptWorkItemHandler.java]: updated

File: exec-workitem/src/main/java/org/jbpm/process/workitem/exec/ExecShellScriptWorkItemHandler.java (hunk 4 above)
Comment (sudhishmk) [reply exec-workitem/src/main/java/org/jbpm/process/workitem/exec/ExecShellScriptWorkItemHandler.java]: updated

File: exec-workitem/src/main/java/org/jbpm/process/workitem/exec/ExecShellScriptWorkItemHandler.java (hunk 5 above)
Comment (sudhishmk) [reply exec-workitem/src/main/java/org/jbpm/process/workitem/exec/ExecShellScriptWorkItemHandler.java:67]: updated

File: exec-workitem/src/main/java/org/jbpm/process/workitem/exec/ExecShellScriptWorkItemHandler.java (hunk 6 above)
Comment (sudhishmk) [reply exec-workitem/src/main/java/org/jbpm/process/workitem/exec/ExecShellScriptWorkItemHandler.java]: updated

File: exec-workitem/src/main/java/org/jbpm/process/workitem/exec/ExecShellScriptWorkItemHandler.java (hunk 1 above)
Comment (tsurdilo) [reply exec-workitem/src/main/java/org/jbpm/process/workitem/exec/ExecShellScriptWorkItemHandler.java:54]: I think you make fair points on possibly having two separate handlers. thanks

File: exec-workitem/src/main/java/org/jbpm/process/workitem/exec/ExecShellScriptWorkItemHandler.java [hunk 11]
```diff
@@ -0,0 +1,162 @@
...
+
+@Wid(widfile = "ExecShellScriptDefinitions.wid", name = "ExecShellScript",
+        displayName = "ExecShellScript",
+        defaultHandler = "mvel: new org.jbpm.process.workitem.exec.ExecShellScriptWorkItemHandler()",
+        documentation = "${artifactId}/index.html",
+        category = "${artifactId}",
+        icon = "ExecShellScript.png",
+        parameters = {
+                @WidParameter(name = "ShellScriptLocation", required = true),
+                @WidParameter(name = "Timeout"),
```
[the remark this answers is not in the document]
Comment (sudhishmk) [reply exec-workitem/src/main/java/org/jbpm/process/workitem/exec/ExecShellScriptWorkItemHandler.java]: Sure, updated to "TimeoutInMillis"

**Comment by tsurdilo** [comment]
@sudhishmk Looks good to me, just wanted to ask for two small things if possible:
1) Code formatting - do you mind formatting the code? 
2) Is it possible to create a test that tests timeout expiring? 

Thanks again for your contribution!

**Comment by tsurdilo** [comment]
I think for formatting you can just follow this from optaplanner project: https://github.com/kiegroup/optaplanner/blob/master/ide-configuration/ide-configuration.adoc

**Comment by sudhishmk** [comment]
> @sudhishmk Looks good to me, just wanted to ask for two small things if possible:
> 
>     1. Code formatting - do you mind formatting the code?
<<    Formatted following the link shared.
>     2. Is it possible to create a test that tests timeout expiring?
<<   Done.

Kindly review the changes.

**Comment by tsurdilo** [comment]
Jenkins retest this

**Comment by sudhishmk** [comment]
@tsurdilo Could you review the latest changes? Thanks

**Comment by tsurdilo** [comment]
@sudhishmk LGTM

@MarianMacik Does this need a QE review or can we go ahead and merge ?

**Review by tsurdilo — APPROVED** [review]

**Comment by sudhishmk** [comment]
@MarianMacik Could you please confirm if this need a QE review?

File: exec-workitem/src/test/java/org/jbpm/process/workitem/exec/ExecShellScriptWorkItemHandlerTest.java [hunk 12]
```diff
@@ -0,0 +1,146 @@
...
+
+        assertNotNull(manager.getResults());
+        assertEquals(1, manager.getResults().size());
+        assertTrue(manager.getResults().containsKey(workItem.getId()));
+
+        Map<String, Object> results = ((TestWorkItemManager) manager).getResults(workItem.getId());
+        List<String> result = (List<String>) results.get(ExecWorkItemHandler.RESULT);
+
+        assertNotNull(result);
+        assertTrue(result.contains("Test Script Started") || result.contains("Test Script Ended"));
```
Comment (MarianMacik) [inline exec-workitem/src/test/java/org/jbpm/process/workitem/exec/ExecShellScriptWorkItemHandlerTest.java:52]: `&&` would be more appropriate here.

File: exec-workitem/src/main/java/org/jbpm/process/workitem/exec/ExecShellScriptWorkItemHandler.java [hunk 13]
```diff
@@ -0,0 +1,159 @@
...
+                if (null != removeEnvironmentVariables && !removeEnvironmentVariables.isEmpty()) {
+                    logger.debug("removeEnvironmentVariables " + removeEnvironmentVariables);
+                    removeEnvironmentVariables.stream().forEach(variable -> envVariables.remove(variable));
+                }
+
+                process = processBuilder.start();
+
+                if (!process.waitFor(timeout, TimeUnit.MILLISECONDS)) {
+                    // Process timed out destroying the process
+                    logger.error("Terminating the process has Process timed out: " + timeout);
```
Comment (MarianMacik) [inline exec-workitem/src/main/java/org/jbpm/process/workitem/exec/ExecShellScriptWorkItemHandler.java:123]: Probably a typo `has` -> `as`.

File: exec-workitem/src/test/java/org/jbpm/process/workitem/exec/ExecShellScriptWorkItemHandlerTest.java (hunk 12 above)
Comment (sudhishmk) [reply exec-workitem/src/test/java/org/jbpm/process/workitem/exec/ExecShellScriptWorkItemHandlerTest.java]: Done

File: exec-workitem/src/main/java/org/jbpm/process/workitem/exec/ExecShellScriptWorkItemHandler.java (hunk 13 above)
Comment (sudhishmk) [reply exec-workitem/src/main/java/org/jbpm/process/workitem/exec/ExecShellScriptWorkItemHandler.java]: Corrected.

Thanks for reviewing.

**Review by MarianMacik — APPROVED** [review]
Thank you, looks good now!

**Comment by MarianMacik** [comment]
jenkins retest this

**Comment by MarianMacik** [comment]
jenkins retest this

**Pull request merged**